clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -271,9 +289,9 @@ public void poll(long timeout) throws IOException {
                 try {
                     /* complete any connections that have finished their handshake */
                     if (key.isConnectable()) {
-                        channel.finishConnect();
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
+                    	// queue this channel for completion after this loop
+                    	connectableChannels.add(channel);
+                    	continue;
```

Comment:
Rats. I missed a line of code. I had a call to channel.prepare() after channel.finishConnect() but it  disappeared. This will handle both SSL handshaking and authentication.